sample: net: sockets: echo_service improvements

ClaCodes opened the pull request:
Build on the echo service sample and improve it in some ways:
- No need for an thread for `accepting`. Instead do it as a socket service. This makes the intention of the socket service clearer in saving threads.
- Handle many connections concurrently

Question:
- <del>Should this be integrated into the original example?
- Maybe in the future: <del>Currently we need to track the open sockets even though the socket service has a copy of the entire list. For two reasons. We need to be able to add/remove sockets, but the api only lets one renew the entire list and if the service is taken down, all the open sockets need to be closed. This could be fixed by modifying the API (for example: `net_socket_service_add`, `net_socket_service_remove`, and some way to close the sockets, when `net_socket_service_unregister` maybe via callback).

rlubos commented:
> Should this be integrated into the original example?

I think that would be the preferred approach, yes.

> for example: net_socket_service_add, net_socket_service_remove

We'd still need to track all the sockets that were passed to the service, so I'm not sure about the benefit of simply refreshing the whole list? But maybe I'm missing something, would need to see more concrete API proposal.

> and some way to close the sockets, when net_socket_service_unregister maybe via callback

I guess we could consider this, but again, it's still the application that owns the sockets (the socket service is just a helper to monitor activity on sockets), so I'm not sure if that'd be more intuitive than the application simpy closing the sockets after deregistering them from the service.

ClaCodes committed:
samples: net: sockets: echo_service: enable multiple tcp clients

Improve the sample to enable multiple tcp clients to connect to the
echo-service.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

ClaCodes committed:
samples: net: sockets: echo_service: Move accept-loop to socket-service

Move the accept-loop into a socket-service. No dedicated thread is needed
just for the accept loop. This better demonstrates the power of the
socket-service in removing the need for dedicated threads. Make some tiny
improvments and necessary changes to the udp-side of the service.

Signed-off-by: Cla Mattia Galliard <clamattia@gmail.com>

rlubos commented on samples/net/sockets/echo_service/src/main.c:
```diff
@@ -25,13 +25,13 @@ LOG_MODULE_REGISTER(net_echo_server_svc_sample, LOG_LEVEL_DBG);
-};
 static struct pollfd sockfd_tcp[MAX_SERVICES];
+static int tcp_socket = -1;
+static int udp_socket = -1;
```
The commit makes no mention of UDP changes, please make them in a separate commit or update the commit description.

rlubos replied:
I was referring to a commit, which from the description claimed to be moving accepting of the TCP connections into the socket service, yet it had unrelated changes to UDP service part of the sample. Just asing to be explicit when describing commits.

rlubos commented on samples/net/sockets/echo_service/src/main.c:
```diff
@@ -227,73 +250,89 @@ static int setup_udp_socket(struct sockaddr_in6 *addr)
-	}
+}
 
+int stop_echo_service(void)
```
This doesn't seem to be used anywhere, perhaps we could use it to restart the services in case of more serious errors (like accept)? Perhaps it'd make sense to keep `main()` for that purpose after all (the resources are allocated in the system anyway, as main shares the thread with late initialization AFAIR).

ClaCodes reviewed (commented):
Thank you for the quick feedback. I tried to address the concerns.

rlubos reviewed (commented):
Please address the compliance issues (too long lines), otherwise LGTM.

ClaCodes commented:
- broken long lines
